### PR TITLE
Adding K8s service name as webhook destination

### PR DESCRIPTION
**Pull request by charliezhangaws**
Signed-off-by: jiahe zhang <zhanncha@amazon.com>

### Description
[Describe what this change achieves]
https://github.com/opensearch-project/notifications/issues/488
### Issues Resolved
[List any issues this PR will resolve]
web-hook URL using K8s service name which normally not a FQDN value
### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt [hunk 1]
```diff
@@ -22,9 +22,52 @@ fun validateEmail(email: String) {
...
+        val out = StringBuilder()
+        val br = java.io.BufferedReader(java.io.InputStreamReader(p.getInputStream()))
+        var line: String? = null
+        var previous: String? = null
+        while (br.readLine().also { line = it } != null) {
+            if (line != previous) {
+                previous = line
+                out.append(line).append('\n')
+            }
+        }
```
Comment (aditjind) [inline notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt:39]: Can you add an example for the FQDN value? Also can we have better variable names? Avoid using `tmp`.

File: notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt [hunk 2]
```diff
@@ -22,9 +22,52 @@ fun validateEmail(email: String) {
...
+            p.destroy()
+        }
+        val tmp: String = out.toString()
+        if (tmp.contains("server can't find")) {
+            println("Not FQDN")
+            return (false) // return false when host is not FQDN
+        } else {
+            println("FQDN")
+            return (true)
+        }
```
Comment (aditjind) [inline notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt:50]: nit: Better to use Ternary operators. Will make the code consice.

**Review by aditjind — DISMISSED** [review]

**Review by aditjind — DISMISSED** [review]

File: notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt [hunk 3]
```diff
@@ -22,9 +22,44 @@ fun validateEmail(email: String) {
     require(isValidEmail(email)) { "Invalid email address" }
 }
 
+fun isFQDN(urlString: String): Boolean {
+    var p: java.lang.Process? = null
+    try {
+        p = java.lang.Runtime.getRuntime()
+            .exec("nslookup " + urlString) // check host is FQDN or not
```
Comment (qreshi) [inline notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt:29]: I've never really seen running a process like this in Java/Kotlin code for validation. I'm not sure if there are any pitfalls to it.

I tried looking up alternatives for FQDN validation and saw the following options:
1. [Libraries like `DomainValidator/URLValidator` exist but it seems they're checking for top-level domains and won't correctly validate FQDNs](https://stackoverflow.com/a/37825902)
2. Regexes seem to get closer but the accepted forms for FQDNs can change and the regex might need to be updated ([this one seemed pretty thorough but it's a pretty big regex so we'd want to add tests to ensure it's doing what we think it is if we decide to add it](https://stackoverflow.com/a/57129482))

@rishabhmaurya @getsaurabh02 Any insights or recommendations on this?

File: notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt [hunk 4]
```diff
@@ -22,9 +22,44 @@ fun validateEmail(email: String) {
...
+                previous = line
+                out.append(line).append('\n')
+            }
+        }
+        if (p.waitFor() == 0) {
+            p.destroy()
+        }
+        return !out.toString().contains("server can't find")
+    } catch (e: java.io.IOException) {
+        println(e.printStackTrace())
```
Comment (qreshi) [inline notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt:45]: Let's change these prints to be logs instead. We could do `warn` level logs here if proceeding with `isFQDN == false` is the expected behavior.

File: notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt [hunk 5]
```diff
@@ -23,8 +33,23 @@ fun validateEmail(email: String) {
...
+        url = URL(urlString)
+    } catch (exception: MalformedURLException) {
+        return false
+    }
+
+    val index: Int = urlString.indexOf("//") + 2
+    val subString: String = urlString.substring(index)
+
+    val regex = Regex(FQDN_REGEX)
+    val isFQDN = regex.matches(subString)
```
Comment (rishabhmaurya) [inline notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt:47]: Lets ask this question on sage.amazon.com 
Also, can we use apache commons [DomainValidator](https://commons.apache.org/proper/commons-validator/apidocs/org/apache/commons/validator/routines/DomainValidator.html) for DomainName validation and maybe use regex for the host name?

File: notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt [hunk 6]
```diff
@@ -23,8 +33,19 @@ fun validateEmail(email: String) {
...
+    if (isFQDN && !DomainValidator.getInstance().isValid(subString)) return false
+    return if (subString == url.host) {
+        (("https" == url.protocol || "http" == url.protocol) && isFQDN) // Support only http/https, other protocols not supported
+    } else {
+        (("https" == url.protocol || "http" == url.protocol) && !isFQDN)
+    }
```
Comment (qreshi) [inline notifications/core-spi/src/main/kotlin/org/opensearch/notifications/spi/utils/ValidationHelpers.kt:48]: This logic seems fairly complex so I'm not sure if there are any edge cases here.

Would it be possible to add more test cases to test all paths in the control flow? For example, is it possible that `DomainValidator.getInstance().isValid` can throw a false negative since it's mostly meant to validate top level domains? Like the [post that was linked earlier](https://stackoverflow.com/questions/37550613/correct-fqdn-uri-invalidate-by-apache-commons-urlvalidator/37825902#37825902) it could be possible that DomainValidator could reject a valid FQDN (`https://sample.url:1234`).

**Review by qreshi — APPROVED** [review]

**Review by rishabhmaurya — APPROVED** [review]
